### PR TITLE
merge latest from upstream 

### PR DESCRIPTION
**Pull request by chrisbc**
This includes
 - our changes from upstream/promote PR which was just merged!
 - other changes from Kevin

### CODE REVIEW

File: src/scratch/kevin/ucerf3/KaikouraRuptureTests.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+import org.opensha.sha.earthquake.faultSysSolution.ruptures.util.SectionDistanceAzimuthCalculator;
+import org.opensha.sha.faultSurface.FaultSection;
+
+import com.google.common.base.Preconditions;
+
+import scratch.UCERF3.FaultSystemRupSet;
+import scratch.UCERF3.inversion.laughTest.PlausibilityResult;
+import scratch.UCERF3.utils.FaultSystemIO;
+
+public class KaikouraRuptureTests {
```
Comment (chrisbc) [inline src/scratch/kevin/ucerf3/KaikouraRuptureTests.java:34]: @voj @mcg-gns  I think we'll ask Kevin to show us this next week - looks interesting!

File: src/scratch/kevin/ucerf3/UniqueRupSetWriter.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import org.opensha.commons.util.FaultUtils;
+import org.opensha.sha.earthquake.faultSysSolution.ruptures.util.UniqueRupture;
+
+import com.google.common.base.Preconditions;
+
+import scratch.UCERF3.FaultSystemRupSet;
+import scratch.UCERF3.FaultSystemSolution;
+import scratch.UCERF3.utils.FaultSystemIO;
+
+public class UniqueRupSetWriter {
```
Comment (chrisbc) [inline src/scratch/kevin/ucerf3/UniqueRupSetWriter.java:20]: It'd be nice to get a few notes on what this utility is doing.

**Review by chrisbc — COMMENTED** [review]
made some notes, but all good AFAIK

**Review by voj — APPROVED** [review]
LGTM. Some tools and classes that I hadn't seen before. Will be good to get some info from Kevin, as @chrisbc said.

**Comment by chrisbc** [comment]
@voj this new code has two compile errors. I think we need to convince Kevin to start doing some CI !

**Pull request merged**